### PR TITLE
Use Rust representation collateral structure to replace C style collateral structure

### PR DESCRIPTION
**Pull request by yao-ji**
To fix issue #310

### CODE REVIEW

**Review by andyzyb — COMMENTED** [review]
Will this break the old customers who are using C style structure?

**Comment by yao-ji** [comment]
> Will this break the old customers who are using C style structure?

Yes. We plan to use a major version upgrade on this change. Old customers can use the old version.

**Comment by haraldh** [comment]
This code is still unsound. After using `convert_collateral()` the original object is not pinned. I see that in my tests, when compiling with `--release`.

**Comment by haraldh** [comment]
This commit https://github.com/haraldh/SGXDataCenterAttestationPrimitives/commit/40bf44e4def88f8433bf7f20a25a98b0bd7c93fd adds some test vectors, where the tests fail with `cargo test --release`.

Same test vectors with my PR from #324 run fine with  `cargo test --release`.

**Comment by haraldh** [comment]
Also I would like to have the Collateral with `pub` elements, so I can serialize them and send them along with the Quote.

**Comment by yao-ji** [comment]
> This code is still unsound. After using `convert_collateral()` the original object is not pinned. I see that in my tests, when compiling with `--release`.

@haraldh `convert_collateral()` is only used for quote verification and we can guarantee that the original collateral is not changed or reallocated in the code. The root cause of failure in release mode is that we use the reference of the returned C-type collateral from `convert_collateral()`, but it is only valid inside the `match` block. We have fixed by moving `convert_collateral()` to outside of the `match` block. This should fix your test. Please try again.

File: QuoteVerification/dcap_quoteverify/sgx-dcap-quoteverify-rs/src/lib.rs [hunk 1]
```diff
@@ -28,24 +28,69 @@
...
-//! Intel(R) Software Guard Extensions Data Center Attestation Primitives (Intel(R) SGX DCAP)
+//! Intel®️ Software Guard Extensions Data Center Attestation Primitives (Intel®️ SGX DCAP)\
 //! Rust wrapper for Quote Verification Library
 //! ================================================
 //!
 //! This is a safe wrapper for **sgx-dcap-quoteverify-sys**.
 
+#[cfg(test)]
+mod test;
+
```
Comment (haraldh) [inline QuoteVerification/dcap_quoteverify/sgx-dcap-quoteverify-rs/src/lib.rs:39]: Maybe you forgot to add these tests.

File: QuoteVerification/dcap_quoteverify/sgx-dcap-quoteverify-rs/src/lib.rs [hunk 2]
```diff
@@ -242,18 +286,17 @@ pub fn tdx_qv_get_quote_supplemental_data_size() -> Result<u32, quote3_error_t>
...
     supplemental_data_size: u32,
     supplemental_data: Option<&mut sgx_ql_qv_supplemental_t>,
 ) -> Result<(u32, sgx_ql_qv_result_t), quote3_error_t> {
-
     let mut collateral_expiration_status = 1u32;
     let mut quote_verification_result = sgx_ql_qv_result_t::SGX_QL_QV_RESULT_UNSPECIFIED;
 
     let p_quote_collateral = match quote_collateral {
-        Some(p) => p,
+        Some(p) => &convert_collateral(p),
```
Comment (haraldh) [inline QuoteVerification/dcap_quoteverify/sgx-dcap-quoteverify-rs/src/lib.rs:299]: still a problem here?

File: QuoteVerification/dcap_quoteverify/sgx-dcap-quoteverify-rs/src/lib.rs [hunk 3]
```diff
@@ -140,18 +185,17 @@ pub fn sgx_qv_get_quote_supplemental_data_size() -> Result<u32, quote3_error_t>
...
```
Comment (haraldh) [inline QuoteVerification/dcap_quoteverify/sgx-dcap-quoteverify-rs/src/lib.rs:198]: still a problem here?

**Review by haraldh — CHANGES_REQUESTED** [review]
`convert_collateral()` still is a pitfall for developers.. better guard it with a PhantomData lifetime as in https://github.com/intel/SGXDataCenterAttestationPrimitives/pull/323

File: QuoteVerification/dcap_quoteverify/sgx-dcap-quoteverify-rs/src/lib.rs (hunk 1 above)
Comment (yao-ji) [reply QuoteVerification/dcap_quoteverify/sgx-dcap-quoteverify-rs/src/lib.rs]: Mis added. We don't have test on this repo.

File: QuoteVerification/dcap_quoteverify/sgx-dcap-quoteverify-rs/src/lib.rs (hunk 2 above)
Comment (yao-ji) [reply QuoteVerification/dcap_quoteverify/sgx-dcap-quoteverify-rs/src/lib.rs]: Changed.

File: QuoteVerification/dcap_quoteverify/sgx-dcap-quoteverify-rs/src/lib.rs (hunk 3 above)
Comment (yao-ji) [reply QuoteVerification/dcap_quoteverify/sgx-dcap-quoteverify-rs/src/lib.rs]: Changed.

**Comment by yao-ji** [comment]
Added the lifetime using PhantomData.

**Review by haraldh — APPROVED** [review]

**Comment by yao-ji** [comment]
Already merged, so closed.

**Pull request closed without merging**